streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java

```diff
@@ -272,6 +274,30 @@ public void handleAssignment(final Map<TaskId, Set<TopicPartition>> activeTasks,
             }
         }
 
+        if (taskCloseExceptions.isEmpty()) {
+            for (final Task oldTask : tasksToRecycle) {
+                final Task newTask;
+                try {
+                    if (oldTask.isActive()) {
+                        final Set<TopicPartition> partitions = standbyTasksToCreate.remove(oldTask.id());
+                        newTask = standbyTaskCreator.createStandbyTaskFromActive((StreamTask) oldTask, partitions);
+                    } else {
+                        final Set<TopicPartition> partitions = activeTasksToCreate.remove(oldTask.id());
+                        newTask = activeTaskCreator.createActiveTaskFromStandby((StandbyTask) oldTask, partitions, mainConsumer);
+                    }
+                    tasks.remove(oldTask.id());
+                    addNewTask(newTask);
+                } catch (final RuntimeException e) {
+                    final String uncleanMessage = String.format("Failed to recycle task %s cleanly. Attempting to close remaining tasks before re-throwing:", oldTask.id());
+                    log.error(uncleanMessage, e);
+                    taskCloseExceptions.put(oldTask.id(), e);
+                    dirtyTasks.add(oldTask);
+                }
+            }
+        } else {
+            dirtyTasks.addAll(tasksToRecycle);
+        }
```

Comment:
This seems like an odd case. Am I right in reading this as, "something went wrong, and we don't know what it was, so we're just going to assume the worst and dump all the tasks that we were hoping to recycle"?

Reply:
Yeah, it still seems to me like if we have to close any tasks as dirty we will ultimately have to do so for them all (as in `handleLostAll`) But that's a big assumption and even if true now, it may not always be...I'll remove this